[commits.webkit.org] Persist with incremental pushes even when they fail

JonWBedard opened the pull request:
#### 830645a9ec7311396e17528331650ccd01b85cad
<pre>
[commits.webkit.org] Persist with incremental pushes even when they fail
<a href="https://bugs.webkit.org/show_bug.cgi?id=291692">https://bugs.webkit.org/show_bug.cgi?id=291692</a>
<a href="https://rdar.apple.com/149489996">rdar://149489996</a>

Reviewed by Brianna Fan.

When an incremental push fails, that&apos;s likely because the remote had more commits than we expected. Rather than
imiediately bail out to a complete push, we should continue attempting incremental pushes, since it&apos;s likely that
the remote is still behind the local.

* Tools/Scripts/libraries/reporelaypy/reporelaypy/checkout.py:
(Checkout.push_update): Continuing attempting to incrementally push, even if an incremental push fails.
(Checkout.forward_update): Return success status of pushes, and log failures.

Canonical link: <a href="https://commits.webkit.org/293849@main">https://commits.webkit.org/293849@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/00d0d23a2d2faa025bd17ce282d84fc7da0cd65b

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/100104 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/19752 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/10048 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/105234 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/50685 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/102145 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/20058 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/28225 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/76202 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/33275 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/103111 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/15325 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/90410 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/56563 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/99577 "Passed tests") | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/15140 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/135/builds/8407 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/50055 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/85052 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/136/builds/8490 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/107593 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/27217 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/19937 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/85157 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/27580 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/86616 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/84692 "Passed tests") | 
| [  ~~🛠 🧪 merge~~](https://ews-build.webkit.org/#/builders/19/builds/21513 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 vision-wk2~~](https://ews-build.webkit.org/#/builders/126/builds/29367 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/7105 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/21069 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/16287 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/27154 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/26965 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/30281 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/28524 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

briannafan commented on Tools/Scripts/libraries/reporelaypy/reporelaypy/checkout.py:
```diff
@@ -249,32 +249,36 @@ def push_update(self, branch=None, tag=None, remote=None, track=False, dest_bran
+                if not did_succeed_this:
+                    sys.stderr.write(f"Failed to push to '{split[0]}'\n")
+                did_succeed_all = did_succeed_all and did_succeed_this
+            return did_succeed_all
```
Is leaving this line in a typo?

JonWBedard replied:
It's not a typo. This _may_ be a bug, it's not obviously so (the early return means we only consume one forwarding rule), but there is an ongoing problem that points to this line possibly being incorrect. I'm still actively investigating.

JonWBedard commented on Tools/Scripts/libraries/reporelaypy/reporelaypy/checkout.py:
```diff
@@ -249,32 +249,36 @@ def push_update(self, branch=None, tag=None, remote=None, track=False, dest_bran
-            return
-        return
+                if not did_succeed_this:
+                    sys.stderr.write(f"Failed to push to '{split[0]}'\n")
```
Removing this logging as I land since it ends up being triggered if we don't push because the local branch is behind the remote. This is often deliberate, and doesn't indicate an error.

JonWBedard committed:
[commits.webkit.org] Persist with incremental pushes even when they fail
https://bugs.webkit.org/show_bug.cgi?id=291692
rdar://149489996

Reviewed by Brianna Fan.

When an incremental push fails, that's likely because the remote had more commits than we expected. Rather than
imiediately bail out to a complete push, we should continue attempting incremental pushes, since it's likely that
the remote is still behind the local.

* Tools/Scripts/libraries/reporelaypy/reporelaypy/checkout.py:
(Checkout.push_update): Continuing attempting to incrementally push, even if an incremental push fails.
(Checkout.forward_update): Return success status of pushes, and log failures.

Canonical link: https://commits.webkit.org/293849@main
